File: src/git.go
package main

import (
	"bytes"
	"os/exec"
	"strings"
)

type GitData struct {
	Head        string `json:"git_head"`
	Summary     string `json:"git_head_summary"`
	ShortHead   string `json:"git_short_head"`
	CommitCount string `json:"git_commit_count"`
}

func Git() (GitData, error) {
	var data GitData
	var err error

	data.Head, err = runGitCommand(`rev-parse HEAD`)
	if err != nil {
		return data, err
	}

	data.Summary, err = runGitCommand(`log -1 --pretty=format:%s HEAD`)
	if err != nil {
		return data, err
	}

	data.ShortHead, err = runGitCommand(`rev-parse --short HEAD`)
	if err != nil {
		return data, err
	}

	data.CommitCount, err = runGitCommand(`rev-list --count master`)
	if err != nil {
		return data, err
	}

	return data, err
}

func runGitCommand(command string) (string, error) {
	var out bytes.Buffer
	cmd := exec.Command("git", strings.Split(command, " ")...)
	cmd.Stdout = &out
	err := cmd.Run()
	output := strings.TrimSuffix(out.String(), "\n")
	return output, err
}

File: src/lib/git.rb
# frozen_string_literal: true

# Functions for working with version control.
module Git
  # Returns the current HEAD
  def self.head
    `git rev-parse HEAD`.chomp
  end

  # Returns the current git HEAD, only the short version.
  def self.short_head
    `git rev-parse --short HEAD`.chomp
  end

  # Returns the commit message summary for the current HEAD.
  def self.head_summary
    `git log -1 --pretty=format:%s #{head}`
  end

  # Returns the number of commits on master.
  def self.commit_count
    `git rev-list --count master`.chomp.to_i
  end

  # Returns all git data for site.
  def self.context
    {
      'git_head' => head,
      'git_head_summary' => head_summary,
      'git_short_head' => short_head,
      'git_commit_count' => commit_count
    }
  end
end
